src/viam/sdk/common/proto_t.hpp

```diff
@@ -0,0 +1,293 @@
+        model_t(T t) : data(std::move(t)) {}
+
+        static void dtor(void* self) noexcept {
+            static_cast<model_t*>(self)->~model_t();
+        }
+
+        static void copy(void const* self, void* dest) {
+            new (dest) model_t(*static_cast<model_t const*>(self));
+        }
+
+        static void move(void* self, void* dest) noexcept {
+            new (dest) model_t(std::move(*static_cast<model_t*>(self)));
+        }
+
+        static void to_proto_value(void const* self, google::protobuf::Value* v);
+
+        static int kind() noexcept;
+
+        static bool equal_to(void const* self, void const* other, const vtable_t& other_vtable) {
+            if (model_t::kind() != other_vtable.kind()) {
+                return false;
+            }
+
+            return *static_cast<T const*>(self) == *static_cast<T const*>(other);
+        }
+
+        static constexpr vtable_t vtable{dtor, copy, move, to_proto_value, kind, equal_to};
+        T data;
+    };
+
+    struct storage_t {
+        static constexpr std::size_t small_size =
+            sizeof(std::unordered_map<std::string, std::string>);
+        template <typename T>
+        storage_t(T t) {
+            static_assert(sizeof(model_t<T>) <= small_size, "Too big!");
+            new (&buf_) model_t<T>(std::move(t));
+        }
+
+        storage_t(const storage_t&) = delete;
+        storage_t(storage_t&&) = delete;
+        storage_t& operator=(const storage_t&) = delete;
+        storage_t& operator=(storage_t&&) = delete;
+
+        storage_t(const storage_t& other, const vtable_t& vtable) {
+            vtable.copy(other.get(), this->get());
+        }
+
+        storage_t(storage_t&& other, const vtable_t& vtable) {
+            vtable.move(other.get(), this->get());
+        }
+
+        void swap(const vtable_t& this_vtable, storage_t& other, const vtable_t& other_vtable) {
+            if (this == &other) {
+                return;
+            }
+
+            unsigned char tmp[small_size];
+            other_vtable.move(other.get(), &tmp);
+            other_vtable.dtor(other.get());
+
+            this_vtable.move(this->get(), other.get());
+            this_vtable.dtor(this->get());
+
+            other_vtable.move(&tmp, this->get());
+            other_vtable.dtor(&tmp);
+        }
+
+        void destruct(const vtable_t& vtable) {
+            vtable.dtor(this->get());
+        }
+
+        template <typename T = void>
+        T* get() {
+            return static_cast<T*>(static_cast<void*>(&buf_));
+        }
+
+        template <typename T = void>
+        T const* get() const {
+            return static_cast<T const*>(static_cast<void const*>(&buf_));
+        }
+
+        unsigned char buf_[small_size];
+    };
+
+    ProtoT(const google::protobuf::Value* value);
+
+    vtable_t vtable_;
+    storage_t self_;
+};
+
+using AttrMap = std::unordered_map<std::string, ProtoT>;
+
+void to_proto_value(std::nullptr_t, google::protobuf::Value* v);
```

Comment:
Should these just be canonically called `to_proto` with the idea that that's the same name we would use elsewhere?

Reply:
I like that, and then the named constructor can be called `from_proto`